### ISSUE TITLE
[security] There is no user provider for user

### ISSUE THREAD

**Issue by thewalkingcoder** [symfony/symfony#35435]
**Symfony version(s) affected**: 4.4.3

**Description**  
After update to 4.4.3 i have a RuntimeException "There is no user provider for user"

**How to reproduce**  
if you have custom provider update to 4.4.3

Sample
```yml
security:
    providers:
        custom-security:
            id: custom.user_provider
    firewalls:
        dev:
            pattern: ^/(_(profiler|wdt)|css|images|js)/
            security: false
        login:
            pattern: ^/login
            security: false
        secured_area:
            pattern: ^/
            guard:
                authenticators:
                    - custom.authentificator
                entry_point: custom.authentificator
            logout:
                path: /logout
```

**Possible Solution**  
Problem is on the file  Symfony/Component/Security/Http/Firewall/ContextListener.php method refreshUser

**Comment by thewalkingcoder** [comment]
After more analyse, my customProvider return bad Class on method supportClass

**Comment by Pi-r-** [comment]
Hi I am having the same issue but with version 3.4.37  [https://stackoverflow.com/questions/59858525/there-is-no-user-provider-for-user-symfony-3-4-37](https://stackoverflow.com/questions/59858525/there-is-no-user-provider-for-user-symfony-3-4-37)

Wen downgrading to 3.4.36 everything works fine.

**Comment by Pi-r-** [comment]
Same thing with version 3.4.37 the issue was from my custom user provider supportsClass methode returning a wrong class

**Comment by ahundiak** [comment]
Just in case anybody is wondering, the 3.4.37 problem was also the result of an invalid supportsClass method.  Looks like the older code was not properly using the method.

**Comment by terox** [comment]
Hello everybody!

I am experiencing the same issue on the Symfony 4.4.3. As all other guys before me, the issue is on the method ```supportsClass``` on my custom user provider. A little explanation about how fix that issue:

Before (fails with the described exception):
```php
public function supportsClass($class)
{
    return $class === UserInterface::class;
}
```
After (works fine):
```php
public function supportsClass($class)
{
    return $class === MyCustomModel::class;
}
```

I was checking the [security component ](https://github.com/symfony/security/) and I think that the issue is introduced in the next commit:
https://github.com/symfony/security/commit/0cfaf60c0666e8967590d0dada80caa7f9b61b02#diff-0cd860ec03e139a658ed3c258b615f21

Sincerely, I think that now it have the expected behaviour; but is breaking bad implemented custom providers. May be some one could correct my words.

Cheers!

**Comment by thewalkingcoder** [comment]
@terox you are right, even if the last implementation is the good way and have sense to fix on 5.x, on 4.4.3 you break semserv.

ping @nicolas-grekas, what do you think ?

**Comment by nicolas-grekas** [comment]
No opinion. Any bug fix is a potential BC break, it's never blackand white. Given all reports mention bad implementations of `supportsClass()`, it looks like fixing then is a good thing.
/cc @linaori WDYT?

**Comment by thewalkingcoder** [comment]
you have right too

**Comment by kissifrot** [comment]
On our side we have the problem when a Doctrine proxy class is used instead of the original one, such as `Proxies\__CG__\App\Entity\OurUser` instead of `App\Entity\OurUser`

**Comment by terox** [comment]
I agree with @nicolas-grekas.

Should the documentation mention or put some example that @kissifrot is commenting? I think that it could be a common issue.

**Comment by kissifrot** [comment]
For most cases to support the proxies you would just have to edit your `supportsClass` like so:
```php
public function supportsClass($class)
{
    return MyUserEntity::class === $class || is_subclass_of($class, MyUserEntity::class);
}
```

**Comment by wouterj** [comment]
To get things clear: What we're seeing here is that `supportsClass()` returns `false` for the currently authenticated user, as it's using some "bad" condition? So if we remove the `supportsClass()` :point_down:  call we end up with a working application?

https://github.com/symfony/symfony/blob/d3942cbe17cb972d50aecfa44936ac36014d592c/src/Symfony/Component/Security/Http/Firewall/ContextListener.php#L178-L180

I think it's good to have this bug fix, but we can do better at reporting the bug. For instance, what about moving this if condition after :point_down: in Symfony 5.x?

https://github.com/symfony/symfony/blob/d3942cbe17cb972d50aecfa44936ac36014d592c/src/Symfony/Component/Security/Http/Firewall/ContextListener.php#L182-L185

We can then improve the exception by saying something like:

 > There is no user provider for user. User provider X was able to refresh the user, but its `supportUser()` method returned false for this class. This will break in Symfony 6.

And then move the if statement to it's current position in Symfony 6, probably speeding up applications with many users providers a bit.

**Comment by linaori** [comment]
@nicolas-grekas this is a use-case that I have discussed with @chalasr I believe, I can't find it right now so must've been on slack as I couldn't see it in the PR or Issue. We considered it to be a bug if your logic in `supportsClass()` was not correct and not a case we should take care of in this bug fix. If this is really causing a lot of issues for people, I suggest we take a look at @wouterj's idea about the exception message.

> To get things clear: What we're seeing here is that `supportsClass()` returns false for the currently authenticated user, as it's using some "bad" condition? So if we remove the `supportsClass()` :point_down: call we end up with a working application?

That depends, it's not clear -especially in older Symfony versions- that the `refreshUser()` should throw an exception for this, as `supportsClass()` has always been there. In most scenarios this worked because the of non-complex authentication processes. 1 user with 1 authenticator and 1 user provider, and you'll not encounter this bug, which makes you believe `supportsClass()` works fine. 

> I think it's good to have this bug fix, but we can do better at reporting the bug. For instance, what about moving this if condition after :point_down: in Symfony 5.x?

This would have to be done in the same in the `ChainUserProvider`. I'm still of opinion that a better check would be something like `supportsUser($user)`, as that's basically what you want to do do. Using the `refreshUser` for this kind of logic, is easy to miss, even without supports (lack of `@throws` in php like java has).

>  can then improve the exception by saying something like [...]

So this suggestion (by moving the logic) would be a DX solution to hint people _what_ is wrong? I would advice to do this in 3.4+ already if it's merely DX as part of this bug fix.

**Comment by MisatoTremor** [comment]
I also ran into this, but think the fix is fine as it is.

I suggest that the upgrade docs for the affected versions should mention this change and the `is_subclass_of` fix as this check seems to me to be the major cause for this issue. Also the default implementation of the `EntityUserProvider::supportsClass` has got this right for years, but generated classes have not (see symfony/maker-bundle#532).

**Comment by chalasr** [comment]
I'm against reverting. It revealed many broken implementations which are likely to open the door for true security holes, especially on 3.4 where changed users are not deauthenticated by default at refresh-time.
PRs welcome for maker-bundle, core documentation and/or DX. But reverting the patch in order to provide better messages (namely upgrade paths) is a no-go to me.

**Comment by chalasr** [comment]
Closing in favor of symfony/maker-bundle#532. Thanks for the report!

**Comment by wouterj** [comment]
@chalasr it's not so much about reverting, it's about improving the error message. Fixing the maker helps future applications, but doesn't help applications upgrading to Symfony 4.4. We should help the upgrading user to easier find this error (as "There is no user provider for this user" isn't a very good indicator to start debugging)

**Comment by linaori** [comment]
> (as "There is no user provider for this user" isn't a very good indicator to start debugging)

What if we change this message to refer to the `supportsUser()` method?

**Comment by chalasr** [comment]
@wouterj right, thanks for clarifying. 👍 on my side

**Comment by wouterj** [comment]
I've implemented @linaori's suggestion in #35472 . I think that's the easiest to not make code much complexer and still give some valuable hints to the user.
